### ISSUE TITLE
Feature Request: ingest ledger backend that connects to stellar-core on port 11625

### ISSUE THREAD

**Issue by leighmcculloch** [stellar/go#3569]
### What problem does your feature solve?

The ingest package is really helpful for building independent ledger ingestion applications, but running a stellar-core or captive core binary complicates deployment. For applications where that complication is too inconvenient the only option is to use checkpoints which limits ledger updates to once every 5 minutes. It'd be awesome if there was a way to use the ingest package with only native Go code and collect ledgers directly from nodes participating in consensus.

### What would you like to see?

I'd like to see a ledger backend implementation that connects to stellar-core via port 11625 and ingests transaction meta, etc updates by participating as a watcher node. The overlay messaging looks relatively simple, performing a simple handshake, using SHA256 HMAC for message auth, and XDR is used for all the messages except the message frame, which is also simple.

### What alternatives are there?

Keep doing what we're doing.

**Comment by ire-and-curses** [comment]
Yes we've had a similar thought before, e.g. https://github.com/stellar/go/issues/2727#issuecomment-648811784

I spoke with @MonsieurNicolas about it. He explained that it was much harder than this (basically not feasible), although I don't recall the details now. I know one challenge is the whole peer discovery and propagation that goes on in the overlay, and since Horizon nodes wouldn’t be real peers, they’d need to be somehow managed differently.

@MonsieurNicolas perhaps can you elaborate better?

**Comment by MonsieurNicolas** [comment]
I don't fully understand what is being discussed here: it *looks* like the goal would be to get full fidelity ledger information without running a core node (checkpoint ledgers are not enough). To do this, you'd need to implement the core protocol in Go.

**Comment by leighmcculloch** [comment]
> the goal would be to get full fidelity ledger information without running a core node

I think this is accurate summary of the goal. At the moment this is only possible with some form of stellar-core running, either in captive core mode or not.

**Comment by leighmcculloch** [comment]
@MonsieurNicolas Is an implementation of the core protocol state machine requires because the nodes only publish transaction sets that mutate the ledger state, and not the actual changes in ledger state?

It sounds like if this was to happen there would be a few components, all are significant in size:
- An overlay implementation in a new package, such as `clients/stellaroverlay`.
- An implementation of the core protocol state machine in a new package? This would probably be quite large?
- An `ingest` ledger backend that uses both of the above two packages.

Is there anything else missing in this list?

**Comment by MonsieurNicolas** [comment]
@leighmcculloch in order to safely apply ledgers based on what you receive from overlay, you would need:
* an implementation of overlay so that you can receive all information needed from the network
* an implementation of SCP (it may be possible to implement a partial version that only relies on `EXTERNALIZE` messages if you're only interested in watcher node behavior) so that you can see which transaction set to apply on top of the previous ledger (to answer your question: yes what we get from consensus is the hash of the previous ledger header + a transaction set to apply on top - not actual changes)
* an implementation of the core protocol so that you can mutate the ledger based on what came out of consensus
    * this also includes an implementation of the bucket list to verify the integrity of the resulting ledger

Basically you're looking at 90% of what core does today.
